src/notebooks/controllers/vscodeNotebookController.ts

```diff
@@ -183,6 +183,13 @@ export class VSCodeNotebookController implements Disposable, IVSCodeNotebookCont
         this.controller.supportedLanguages = this.languageService.getSupportedLanguages(kernelConnection);
         // Hook up to see when this NotebookController is selected by the UI
         this.controller.onDidChangeSelectedNotebooks(this.onDidChangeSelectedNotebooks, this, this.disposables);
+        this.notebookApi.onDidCloseNotebookDocument(
```

Comment:
Associated documents were surviving tests closing notebooks.

Reply:
Which could also potentially happen in the real product.

User opens notebook
Connects to remote
User closes notebook
User opens another notebook
Connects to local
Kernel list still includes remote connection from notebook that isn't open anymore

Reply:
New test I added is verifying this isn't the case

Reply:
This might improve performance and potentially fix other unnoticed bugs! I think? This looks really promising to me